### DISCUSSION TITLE
What are the various files in the `queries` directories of language grammars?

### DISCUSSION THREAD

**Discussion by patrickdevivo** [tree-sitter/tree-sitter#1251] [Q&A] [answered]
Hi folks!

I apologize if this is covered elsewhere - I'm still fairly new to Tree-sitter and just getting my bearings 😃 . I've been poking around some of the grammar repositories, such as for [go](https://github.com/tree-sitter/tree-sitter-go) and [javascript](https://github.com/tree-sitter/tree-sitter-javascript) and I notice in many (all?) of these repos there's a `queries` directory that holds some files, I think always a `highlights.scm` and sometimes others (`tags.scm`, `locals.scm`) and I'm trying to understand what the differences are in these queries (the files) in how they're used (and what the file names typically mean). I assume `highlights.scm` is the main syntax-highlighting query use case, but it's not immediately clear to me what some of the other query files are for.

I'm asking as I'm very interested in using Tree-sitter in a project, not for syntax-highlighting, but as a way to query/ask questions about parsed source code: "how many functions are in this file?" "what are their names?" "what are their signatures?" "what are the imported libraries?" etc

Thanks!
Patrick

**Comment by sogaiu** [comment]
Just to touch on `locals.scm` -- may be it's related to: https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables

**Comment by ahlinc** [comment, marked as answer]
- `tree-sitter highlight` command uses:
  - [`tree-sitter-highlight`](https://github.com/tree-sitter/tree-sitter/tree/master/highlight#readme) Rust library that uses:
    - `highlights.scm` - to markup a language symbols for highlighting.
    - `injections.scm` - [to](https://github.com/tree-sitter/tree-sitter/blob/0926fad1e2fdcd86d395934b55b7c4c8112c013b/highlight/src/lib.rs#L219) [define](https://github.com/tree-sitter/tree-sitter/blob/0926fad1e2fdcd86d395934b55b7c4c8112c013b/highlight/src/lib.rs#L253-L254) scopes of embedded languages and their types that are used to simultaneously highlight the main and the embedded languages.
    - `locals.scm` - to [define](https://github.com/tree-sitter/tree-sitter/blob/0926fad1e2fdcd86d395934b55b7c4c8112c013b/highlight/src/lib.rs#L255-L258) matching for a file local definitions and definition references and also scopes for complex constructs for ex. function or class bodies. It's accompanying file for the `highlights.scm` and @sogaiu provided the good docs link.
- `tree-sitter tags` command computes code navigation tags and uses:
  - [`tree-sitter-tags`](https://github.com/tree-sitter/tree-sitter/tree/master/tags#readme) Rust library that uses:
    - `tags.scm` - [to](https://github.com/tree-sitter/tree-sitter/blob/0926fad1e2fdcd86d395934b55b7c4c8112c013b/tags/src/lib.rs#L140-L145) [markup](https://github.com/tree-sitter/tree-sitter/blob/0926fad1e2fdcd86d395934b55b7c4c8112c013b/tags/src/lib.rs#L149-L153) [local](https://github.com/tree-sitter/tree-sitter/blob/0926fad1e2fdcd86d395934b55b7c4c8112c013b/tags/src/lib.rs#L185) [to](https://github.com/tree-sitter/tree-sitter/blob/0926fad1e2fdcd86d395934b55b7c4c8112c013b/tags/src/lib.rs#L191) a file symbol definitions and symbol references.
    - `locals.scm` - for the same purpose as above.

You may find other files in the `queries` directory but that are something Github specific stuff.

### Edit 1
**It is good to note** that both `tree-sitter-highlight` and `tree-sitter-tags` libraries are usable not only for Rust code bases, these libraries also exposes C APIs and can be binary linked to any other languages as binary libraries where there is possibility to bind to C API compatible libraries.

**Comment by razzeee** [comment]
Regarding your question, you can use queries to tackle that, have a look at
the docs on how to write them.
